presto-hive/src/main/java/io/prestosql/plugin/hive/metastore/thrift/ThriftHiveMetastore.java

```diff
@@ -622,6 +636,56 @@ private void deletePartitionColumnStatistics(HiveIdentity identity, String datab
         }
     }
 
+    private void setColumnStatistics(HiveIdentity identity, String objectName, List<ColumnStatisticsObj> statistics, Call1<List<ColumnStatisticsObj>> saveColumnStatistics)
+            throws TException
+    {
+        boolean containsDateStatistics = statistics.stream().anyMatch(stats -> stats.getStatsData().isSetDateStats());
+
+        Optional<Boolean> metastoreSupportsDateStatistics = this.metastoreSupportsDateStatistics.get();
+        if (containsDateStatistics && metastoreSupportsDateStatistics.equals(Optional.of(FALSE))) {
+            log.debug("Skipping date statistics for %s because metastore does not support them", objectName);
+            statistics = statistics.stream()
+                    .filter(stats -> !stats.getStatsData().isSetDateStats())
+                    .collect(toImmutableList());
+            containsDateStatistics = false;
+        }
+
+        if (!containsDateStatistics || metastoreSupportsDateStatistics.equals(Optional.of(TRUE))) {
+            try (ThriftMetastoreClient client = createMetastoreClient(identity)) {
+                saveColumnStatistics.call(client, statistics);
+            }
+            return;
+        }
+
+        List<ColumnStatisticsObj> statisticsExceptDate = statistics.stream()
+                .filter(stats -> !stats.getStatsData().isSetDateStats())
+                .collect(toImmutableList());
+
+        List<ColumnStatisticsObj> dateStatistics = statistics.stream()
+                .filter(stats -> stats.getStatsData().isSetDateStats())
+                .collect(toImmutableList());
+
+        verify(!dateStatistics.isEmpty() && metastoreSupportsDateStatistics.equals(Optional.empty()));
+
+        try (ThriftMetastoreClient client = createMetastoreClient(identity)) {
+            saveColumnStatistics.call(client, statisticsExceptDate);
+
+            try {
+                saveColumnStatistics.call(client, dateStatistics);
+            }
+            catch (TException e) {
+                // When `dateStatistics.size() > 1` we expect something like "TApplicationException: Required field 'colName' is unset! Struct:ColumnStatisticsObj(colName:null, colType:null, statsData:null)"
+                // When `dateStatistics.size() == 1` we expect something like "TTransportException: java.net.SocketTimeoutException: Read timed out"
```

Comment:
`dateStatistics.size() == 1` ->  Read timed out 😱 

I think we might want to understand why it blocks. Maybe it deadlocks or thread somehow got blocked. And so calling this several times could consume all HMS threads causing HMS to become unresponsive.

Reply:
It blocks because Hive / thrift lib on the receiver doesn't do a great job.
I can't handle this better.

Reply:
I mean that in case of timeout maybe we should not retry any more.

Reply:
retry is time-bound too, so I dont think anything more is necessary here